cpg_pipes/providers/storage.py

```diff
@@ -0,0 +1,138 @@
+"""
+Abstract storage provider.
+"""
+
+import pathlib
+from enum import Enum
+from abc import ABC, abstractmethod
+from typing import Union
+
+from cloudpathlib import CloudPath
+from cloudpathlib.anypath import to_anypath
+
+
+# Path can be either a cloud URL or a local posix file path.
+Path = Union[CloudPath, pathlib.Path]
+
+# Using convenience method from cloudpathlib to parse a path string.
+to_path = to_anypath
+
+
+class Namespace(Enum):
+    """
+    Storage namespace.
+    https://github.com/populationgenomics/team-docs/tree/main/storage_policies#main-vs-test
+    """
+
+    MAIN = 'main'
+    TEST = 'test'
+
+
+class Cloud(Enum):
+    """
+    Cloud storage provider and correponding protocol prefix.
+    """
+
+    GS = 'gs'
+    AZ = 'az'
+
+
+class StorageProvider(ABC):
+    """
+    Abstract class for implementing storage bucket policy.
+    Onty get_bucket() method is required, however other methods
+    are available to override as well.
+    """
+
+    def __init__(self, cloud: Cloud):
+        self.cloud = cloud
+
+    @abstractmethod
+    def get_bucket(
+        self,
+        dataset: str,
+        namespace: Namespace,
+        suffix: str | None = None,
+        version: str | None = None,
+        sample: str = None,
+    ) -> Path:
+        """
+        Bucket to write primary results.
+        @param dataset: dataset/stack name
+        @param namespace: namespace (test or main)
+        @param suffix: (optional) suffix to append to the bucket name
+        @param version: (optional) pipeline version
+        @param sample: (optional) sample name
+        """
+
+    def get_analysis_bucket(
```

Comment:
Does this refer to our `main` buckets? Why isn't this implemented in the CPG provider?

Reply:
Other implementations might also use some sort of an analysis bucket. Or they can disregard the `suffix` parameter. But `get_analysis_bucket` must exist in the abstract class anyway, because the stages don't have to know about the specific implementations.

Reply:
Right, but we don't have `analysis` buckets anymore?

Reply:
But we do, we re-added analysis buckets at some point:
https://github.com/populationgenomics/analysis-runner/blob/main/stack/__main__.py#L236-L238
Though perhaps the goal was to allow read by human, and we allow that for main buckets already, so we perhaps don't need analysis buckets anymore? And I can see that we don't have them described in team-docs.

Reply:
Thanks for pointing that out! I had completely forgotten that we revived the `analysis` buckets, even though I wrote that PR in September myself ;).

I've checked the permissions again to understand the difference between the `analysis` and `main` buckets in the current setup. As far as I can tell, members of the access group currently only have *listing* permissions for the `main` bucket, but can *read* the contents for the `analysis` bucket:
https://github.com/populationgenomics/analysis-runner/blob/f6045469308d1445cc262b5a1ecd5cbcafc88ab0/stack/__main__.py#L544-L549

I'll now finally update the storage policies doc (which got me confused in the first place).